File: app/models/conversation.py
"""对话数据模型 - 对应原项目 ChatTypeHistory / ChatHistoryMapper"""

from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, JSON
from sqlalchemy.orm import relationship
from app.core.database import Base


class Conversation(Base):
    """对话表"""
    __tablename__ = "conversations"

    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(200), default="新对话")
    chat_type = Column(String(20), default="assistant")  # assistant / rag / agent
    pinned = Column(Integer, default=0)  # 0=未置顶, 1=置顶
    created_at = Column(DateTime, default=datetime.now)
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now)

    messages = relationship("Message", back_populates="conversation", cascade="all, delete-orphan")


class Message(Base):
    """消息表"""
    __tablename__ = "messages"

    id = Column(Integer, primary_key=True, autoincrement=True)
    conversation_id = Column(Integer, ForeignKey("conversations.id"), nullable=False)
    role = Column(String(20), nullable=False)  # user / assistant / system
    content = Column(Text, nullable=False)
    sources = Column(JSON, nullable=True)  # RAG 引用来源
    steps = Column(JSON, nullable=True)  # Agent 推理步骤 [{step, action, content}]
    token_usage = Column(JSON, nullable=True)  # {"prompt_tokens": N, "completion_tokens": N, "total_tokens": N}
    created_at = Column(DateTime, default=datetime.now)

    conversation = relationship("Conversation", back_populates="messages")

File: frontend/src/api/conversation.ts
import api from './index'

export interface Conversation {
  id: number
  title: string
  chat_type: string
  pinned: boolean
  created_at: string
}

export interface Message {
  id?: number
  role: 'user' | 'assistant'
  content: string
  sources?: Source[]
  steps?: Step[]
  token_usage?: TokenUsage
  error?: string
}

export interface Source {
  doc: string
  excerpt: string
}

export interface Step {
  step: number
  action: 'reasoning' | 'tool_start' | 'tool_end' | 'node_enter'
  content: string
  tool?: string
}

export interface TokenUsage {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
}

/** 获取对话列表 */
export async function getConversations(): Promise<Conversation[]> {
  const { data } = await api.get('/conversations')
  return data
}

/** 获取对话消息 */
export async function getMessages(conversationId: number): Promise<Message[]> {
  const { data } = await api.get(`/conversations/${conversationId}/messages`)
  return data
}

/** 删除对话 */
export async function deleteConversation(conversationId: number) {
  await api.delete(`/conversations/${conversationId}`)
}

/** 更新标题 */
export async function updateTitle(conversationId: number, title: string) {
  await api.put(`/conversations/${conversationId}/title`, { title })
}

/** 切换置顶 */
export async function togglePin(conversationId: number) {
  const { data } = await api.post(`/conversations/${conversationId}/pin`)
  return data
}

/** 生成标题 */
export async function generateTitle(conversationId: number) {
  await api.post(`/conversations/${conversationId}/generate-title`)
}
